lib/core/checker/type_expression_checker.cc

```diff
@@ -0,0 +1,222 @@
+#include "type_expression_checker.h"
+
+#include "core/ast/ast.h"
+#include "core/ast/expression.h"
+#include "location.hh"
+
+#include <string>
+#include <vector>
+
+namespace dbuf::checker {
+
+TypeExpressionChecker::TypeExpressionChecker(ast::AST &ast)
+    : ast_(ast) {}
+
+void TypeExpressionChecker::GetConstructorToEnum() {
+  for (const auto &ast_enum : ast_.enums) {
+    for (const auto &rule : ast_enum.second.pattern_mapping) {
+      for (const auto &constructor : rule.outputs) {
+        constructor_to_enum_[constructor.identifier.name] = ast_enum.second.identifier.name;
+      }
+    }
+  }
+}
+
+void TypeExpressionChecker::PushScope() {
```

Comment:
It bothers me a little bit that definition order doesn't match the declaration order :D